go/interpreter/lfvm/interpreter_intergration_test.go

```diff
@@ -0,0 +1,689 @@
+				if want, got := test.gas.after, ctxt.gas; want != got {
+					t.Errorf("execution failed: gas consumption is %v, wanted %v",
+						test.gas.before-got,
+						test.gas.before-want)
+				}
+
+				// Check gas refund
+				if want, got := test.gasRefund, ctxt.refund; want != got {
+					t.Errorf("execution failed: gas refund is %v, wanted %v", got, want)
+				}
+			})
+		})
+	}
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Helper utils
+
+// testParameter is a tool that allows to define before-after pairs of values for a test.
+type testParameter[T any] struct {
+	before, after T
+}
+
+type gas tosca.Gas
+type pcOverride int32
+
+// overridePc will mark the tests to check for a specific PC value.
+// When not used, len(code) is used as the expected PC value.
+func overridePc(pc int32) *pcOverride {
+	res := pcOverride(pc)
+	return &res
+}
+
+// revisionConstraint is a helper struct to define a revision constraint for a test.
+// revisions constraints are defined by the [min, max) range.
+// zero value means no constraint.
+type revisionConstraint struct {
+	min, max *tosca.Revision
+}
+
+// isValidFor checks if the constraint is valid for revision.
+func (rc revisionConstraint) isValidFor(revision tosca.Revision) bool {
+	if rc.min != nil && revision < *rc.min {
+		return false
+	}
+	if rc.max != nil && revision >= *rc.max {
+		return false
+	}
+	return true
+}
+
+// validBefore creates a revision constraint that is valid for revisions greater equal than argument.
+func validFrom(revision tosca.Revision) revisionConstraint {
+	return revisionConstraint{min: &revision}
+}
+
+// validBefore creates a revision constraint that is valid for revisions older than argument.
+func validBefore(revision tosca.Revision) revisionConstraint {
+	return revisionConstraint{max: &revision}
+}
+
+// validInRange creates a revision constraint that is valid only for revision.
+func validOnlyIn(revision tosca.Revision) revisionConstraint {
+	next := revision + 1
+	return revisionConstraint{min: &revision, max: &next}
+}
+
+// validInRange creates a revision constraint that is valid in the range [min, max]
+func validInRange(min, max tosca.Revision) revisionConstraint {
+	next := max + 1
+	return revisionConstraint{min: &min, max: &next}
+}
+
+// generateOpCodesInRange generates a slice of opcodes in the range [start, end].
+func getOpCodesInRange(start, end OpCode) []OpCode {
+	opCodes := make([]OpCode, end-start+1)
+	for i := start; i <= end; i++ {
+		opCodes[i-start] = i
+	}
+	return opCodes
+}
+
+func generateTestCaseFor(f func(op OpCode) testDeclaration, opCodes []OpCode) []testDeclaration {
+	tests := make([]testDeclaration, len(opCodes))
+	for i, op := range opCodes {
+		tests[i] = f(op)
+	}
+	return tests
+}
+
+func toKey(value byte) tosca.Key {
+	res := tosca.Key{}
+	res[len(res)-1] = value
+	return res
+}
+
+func toWord(value int) tosca.Word {
+	v := uint256.NewInt(uint64(value))
+	return v.Bytes32()
+}
```

Comment:
idea for another PR: move this into tosca package if there are similar functions in other packages to prevent duplicated functionality